### PR TITLE
fix: preserve partition topology when counting Spark rows

### LINKED ISSUE #2278: [BUG]java.lang.ArrayIndexOutOfBoundsException on multi-node cluster run
### SynapseML version

com.microsoft.azure:synapseml_2.12:0.11.4-spark3.3

### System information

- **Language version** (e.g. python 3.8, scala 2.12): python 3.9
- **Spark Version** (e.g. 3.2.3): 3.3.2
- **Spark Platform** (e.g. Synapse, Databricks): Databricks

### Describe the problem

I have a for-loop lightgbm fit job for rolling back validation;
The job failed on multi-node cluster with log error `Connection Refused`, and after checked the failed tasks, the executor failed with detail error message `java.lang.ArrayIndexOutOfBoundsException` and caused the `Connection Refused` error;

Meanwhile the job can run on single-node cluster without any issue.

The dataframe sent to model is around 48,000, with partition as below

Partition 0 has 19000 records
Partition 1 has 18000 records
Partition 2 has 7000 records
Partition 3 has 4000  records

And the issue cannot be fixed by `df.repartition(5)`.

Screenshot 2024-09-04 at 21 16 29

### Code to reproduce issue

```python #
max_base_date = '2024-09-01'
tmp_train_df = train_merged_df.where(sf.col('base_date')<max_base_date).cache()
tmp_actual_df = actual_merged_df.where(sf.col('base_date')<max_base_date).cache()
model.fit(tmp_train_df, tmp_actual_df)
```

### Other info / logs

_No response_

### What component(s) does this bug affect?

- [ ] `area/cognitive`: Cognitive project
- [ ] `area/core`: Core project
- [ ] `area/deep-learning`: DeepLearning project
- [ ] `area/lightgbm`: Lightgbm project
- [ ] `area/opencv`: Opencv project
- [ ] `area/vw`: VW project
- [ ] `area/website`: Website
- [ ] `area/build`: Project build system
- [ ] `area/notebooks`: Samples under notebooks folder
- [ ] `area/docker`: Docker usage
- [ ] `area/models`: models related issue

### What language(s) does this bug affect?

- [ ] `language/scala`: Scala source code
- [X] `language/python`: Pyspark APIs
- [ ] `language/r`: R APIs
- [ ] `language/csharp`: .NET APIs
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/synapse`: Azure Synapse integrations
- [ ] `integrations/azureml`: Azure ML integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by ranadeepsingh**
## Summary

Fix per-partition row counting so it preserves the exact partition topology used by downstream LightGBM training tasks.

Fixes #2278.

This replaces only the proven intent behind #2282; it does not reuse that PR's corrupted diff or modify the old PR.

## Root cause and primary failure

`ClusterUtil.getNumRowsPerPartition` counted a separate `df.select(lit(0))` query. With adaptive execution enabled, Spark can coalesce that cheap projected query differently from the real training DataFrame.

A current-master AQE fixture produced:

- real DataFrame: 20 partitions
- projected counting query: 4 partitions
- real task IDs indexing the shortened array: primary `ArrayIndexOutOfBoundsException`, including `Index 18 out of bounds for length 4`

That is the same partition-ID-shaped failure reported in #2278. Once a worker dies on this index access, the remaining LightGBM workers report secondary connection failures.

## Change

Count on the original DataFrame RDD rather than constructing a separately optimized literal projection. This intentionally trades projection pruning for the required topology correctness.

The regression test:

1. creates a 20-partition shuffle whose full rows remain above AQE's coalescing target;
2. proves the old literal projection coalesces to fewer partitions; and
3. verifies returned counts exactly match every real DataFrame partition.

## Separately investigated hypotheses

- **Feature width:** malformed short dense rows can independently fail in `SampledData.pushRow`, but no evidence ties inconsistent feature widths to #2278, so no unrelated validation change is included.
- **Barrier/network startup:** the existing barrier `TrainValidationSplit` test passes; connection errors follow the primary task failure rather than precede it.
- **Native pointer lifecycle:** the 300-iteration custom-objective path remains functional. The cleanup proposed in #2282 cannot explain a partition-index exception, and no stable native-memory regression was established, so it is not included.

## Validation

- `core/testOnly com.microsoft.azure.synapse.ml.core.utils.VerifyClusterUtil`
- `core/compile`, `core/Test/compile`
- `core/scalastyle`, `core/Test/scalastyle`
- `lightgbm/Test/compile`
- `lightgbm/scalastyle`, `lightgbm/Test/scalastyle`
- LightGBM streaming barrier `TrainValidationSplit` test
- LightGBM streaming custom-loss test (300 iterations)

## Review

- No public method signatures changed.
- No serialization, authentication, network, file, or secret-handling behavior changed.
- The test uses an isolated Spark session and adds no external service dependency.

### CODE REVIEW

File: core/src/main/scala/com/microsoft/azure/synapse/ml/core/utils/ClusterUtil.scala [hunk 1]
```diff
@@ -44,9 +43,9 @@ object ClusterUtil {
...
     */
   def getNumRowsPerPartition(df: DataFrame, labelCol: Column): Array[Long] = {
-    val indexedRowCounts: Array[(Int, Long)] = df
-      .select(typedLit(0.toByte))
-      .rdd
+    // Use the DataFrame's own RDD so adaptive execution cannot produce a different
+    // partition topology for a projected counting query.
+    val indexedRowCounts: Array[(Int, Long)] = df.rdd
       .mapPartitionsWithIndex({case (i,rows) => Iterator((i,rows.size.toLong))}, true)
       .collect()
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply core/src/main/scala/com/microsoft/azure/synapse/ml/core/utils/ClusterUtil.scala:51]: Addressed in 1ce1ec4c43: the Scaladoc now explicitly records that labelCol is retained for API compatibility and must not be projected because that can change the AQE topology. The old implementation also selected a literal rather than using labelCol; renaming the public parameter would risk named-argument source compatibility.

File: core/src/test/scala/com/microsoft/azure/synapse/ml/core/utils/VerifyClusterUtil.scala [hunk 2]
```diff
@@ -15,4 +16,38 @@ class VerifyClusterUtil extends TestBase {
     assert(ClusterUtil.getDefaultNumExecutorCores(spark, log, Option("spark://localhost:7077")) ==
       ClusterUtil.getJVMCPUs(spark))
   }
+
+  test("Verify row counts preserve the DataFrame partition topology") {
+    val adaptiveSpark = spark.newSession()
+    val partitionCount = 20
+    adaptiveSpark.conf.set("spark.sql.adaptive.enabled", value = true)
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply core/src/test/scala/com/microsoft/azure/synapse/ml/core/utils/VerifyClusterUtil.scala:24]: The session is intentionally not closed: SparkSession.newSession shares TestBase's SparkContext, and SparkSession.close/stop would stop that shared context and break the rest of the suite. Its SQLConf is session-local and the session object becomes collectible after the test. I added an explicit lifecycle comment in 1ce1ec4c43.

**Review by BrendanWalsh — APPROVED** [review]
Reviewed the AQE partition-topology change and regression coverage. No blocking issues found.
